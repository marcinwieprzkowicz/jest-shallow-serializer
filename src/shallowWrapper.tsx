import React, { type JSX, type PropsWithChildren } from "react";
import { jest } from "@jest/globals";
import set from "set-value";
import stringify from "json-stringify-safe";
import type { ShallowWrapper } from "./types";
import { getChildren } from "./utils/getChildren";
import { getFirstTag } from "./utils/getFirstTag";
import { jsonReplacer } from "./utils/jsonReplacer";

export interface MockOptions {
  renderAs: string;
}

export const shallowWrapper = (
  modulePath: string,
  ...componentNames: string[]
) => {
  return (): ShallowWrapper => {
    const actual = jest.requireActual<Record<string, unknown>>(modulePath);
    const mocked = new Map();

    const wrapper: ShallowWrapper = {
      ...actual,
      mock: (name: string, options?: MockOptions) => {
        mocked.set(name, options);
      },
      unmock: (name: string) => {
        mocked.delete(name);
      },
    };

    componentNames.forEach((componentName) => {
      set(
        wrapper,
        componentName,
        jest.fn(({ children, ...props }: PropsWithChildren) => {
          const Component = actual[
            componentName
          ] as ({}: PropsWithChildren) => JSX.Element;

          if (mocked.has(componentName)) {
            const options = mocked.get(componentName);
            const tag =
              options?.renderAs || getFirstTag(require.resolve(modulePath));

            return React.createElement(
              tag,
              {
                "data-shallow-name": componentName,
                "data-shallow-props": stringify(
                  props,
                  jsonReplacer,
                  undefined,
                  () => "[Circular]"
                ),
              },
              getChildren(children)
            );
          }

          return <Component {...props}>{children}</Component>;
        })
      );
    });

    return wrapper;
  };
};

declare global {
  function shallowWrapper(
    modulePath: string,
    ...componentName: string[]
  ): () => ShallowWrapper;
}
