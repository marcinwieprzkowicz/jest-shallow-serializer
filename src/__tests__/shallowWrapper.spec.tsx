import React, { type JSX, type ComponentType, lazy, act } from "react";
import { render } from "@testing-library/react";
import { shallowWrapper } from "../shallowWrapper";
import type { TestComponentProps } from "./__fixtures__/TestComponent";
import type { TestComponentSuspenseProps } from "./__fixtures__/TestComponentSuspense";
import type { TestComponentContextProps } from "./__fixtures__/TestComponentContext";

type Component = (props: TestComponentProps) => JSX.Element;
type ComponentTd = () => JSX.Element;
type ComponentSuspense = (props: TestComponentSuspenseProps) => JSX.Element;
type ComponentContext = (props: TestComponentContextProps) => JSX.Element;

let objectWithCircularReference = {};
Object.assign(objectWithCircularReference, {
  nested: {
    arr: [
      {
        item: {
          value: "lorem ipsum",
        },
      },
    ],
  },
});
Object.assign(objectWithCircularReference, {
  circularReference: objectWithCircularReference,
});

describe("shallowWrapper", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("returns wrapped component", () => {
    const wrapper = shallowWrapper(
      "./__tests__/__fixtures__/TestComponent.tsx",
      "TestComponent"
    )();

    expect(wrapper.mock).toBeDefined();
    expect(wrapper.unmock).toBeDefined();
    expect(wrapper.TestComponent).toBeDefined();
  });

  it("does not mock `component` by default", () => {
    const wrapper = shallowWrapper(
      "./__tests__/__fixtures__/TestComponent.tsx",
      "TestComponent"
    )();
    const TestComponent = wrapper.TestComponent as Component;

    const { asFragment } = render(
      <TestComponent prop="lorem ipsum">children</TestComponent>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  describe("`mock` function", () => {
    it("mocks `component`", () => {
      const wrapper = shallowWrapper(
        "./__tests__/__fixtures__/TestComponent.tsx",
        "TestComponent"
      )();
      wrapper.mock("TestComponent");

      const TestComponent = wrapper.TestComponent as Component;
      const { asFragment } = render(
        <TestComponent prop="lorem ipsum">children</TestComponent>
      );

      expect(asFragment()).toMatchSnapshot();
      wrapper.unmock("TestComponent");
    });

    it("supports circular references", () => {
      const wrapper = shallowWrapper(
        "./__tests__/__fixtures__/TestComponent.tsx",
        "TestComponent"
      )();
      wrapper.mock("TestComponent");

      const TestComponent = wrapper.TestComponent as Component;
      const { asFragment } = render(
        <TestComponent prop="lorem ipsum" obj={objectWithCircularReference}>
          children
        </TestComponent>
      );

      expect(asFragment()).toMatchSnapshot();
      wrapper.unmock("TestComponent");
    });

    it("supports functions", () => {
      const wrapper = shallowWrapper(
        "./__tests__/__fixtures__/TestComponent.tsx",
        "TestComponent"
      )();
      wrapper.mock("TestComponent");

      const TestComponent = wrapper.TestComponent as Component;
      const { asFragment } = render(
        <TestComponent
          prop="lorem ipsum"
          obj={{ method: () => {}, mockMethod: jest.fn() }}
        >
          children
        </TestComponent>
      );

      expect(asFragment()).toMatchSnapshot();
      wrapper.unmock("TestComponent");
    });

    it("supports nested component paths", () => {
      const wrapper = shallowWrapper(
        "./__tests__/__fixtures__/TestComponentContext",
        "TestContext.Provider"
      )();
      wrapper.mock("TestContext.Provider");

      const TestComponentContext =
        wrapper.TestComponentContext as ComponentContext;
      const { asFragment } = render(
        <TestComponentContext>lorem ipsum</TestComponentContext>
      );

      expect(asFragment()).toMatchSnapshot();
      wrapper.unmock("TestContext.Provider");
    });

    describe("lazy components", () => {
      let wrapper;

      beforeEach(() => {
        wrapper = shallowWrapper(
          "./__tests__/__fixtures__/TestComponentSuspense.tsx",
          "TestComponentSuspense"
        )();
        wrapper.mock("TestComponentSuspense");
      });

      afterEach(() => {
        wrapper.unmock("TestComponentSuspense");
      });

      it("renders resolved components", async () => {
        const promise = Promise.resolve<{ default: ComponentType<{}> }>({
          default: () => <div>I'm lazy</div>,
        });
        const LazyComponent = lazy(() => promise);

        const TestComponentSuspense =
          wrapper.TestComponentSuspense as ComponentSuspense;

        const { asFragment } = render(
          <TestComponentSuspense>
            <h1>Heading</h1>
            <LazyComponent />
          </TestComponentSuspense>
        );

        await act(() => promise);
        expect(asFragment()).toMatchSnapshot();
      });

      it("does not render unresolved components", async () => {
        const promise = Promise.resolve<{ default: ComponentType<{}> }>({
          default: () => <div>I'm lazy</div>,
        });
        const LazyComponent = lazy(() => promise);

        const TestComponentSuspense =
          wrapper.TestComponentSuspense as ComponentSuspense;

        const { asFragment } = render(
          <TestComponentSuspense>
            <h1>Heading</h1>
            <LazyComponent />
          </TestComponentSuspense>
        );

        expect(asFragment()).toMatchSnapshot();
        await act(() => promise);
      });
    });

    it("uses the first tag from the component body", () => {
      const wrapper = shallowWrapper(
        "./__tests__/__fixtures__/TestComponentTd.tsx",
        "TestComponentTd"
      )();
      wrapper.mock("TestComponentTd");

      const TestComponentTd = wrapper.TestComponentTd as ComponentTd;
      const { asFragment } = render(<TestComponentTd />, {
        wrapper: ({ children }) => (
          <table>
            <tbody>
              <tr>{children}</tr>
            </tbody>
          </table>
        ),
      });

      expect(asFragment()).toMatchSnapshot();
      wrapper.unmock("TestComponentTd");
    });

    it("uses `options.renderAs` as render tag", () => {
      const wrapper = shallowWrapper(
        "./__tests__/__fixtures__/TestComponent.tsx",
        "TestComponent"
      )();
      wrapper.mock("TestComponent", { renderAs: "span" });

      const TestComponent = wrapper.TestComponent as Component;
      const { asFragment } = render(
        <TestComponent prop="lorem ipsum">children</TestComponent>
      );

      expect(asFragment()).toMatchSnapshot();
      wrapper.unmock("TestComponent");
    });
  });

  describe("`unmock` function", () => {
    it("unmocks `component`", () => {
      const wrapper = shallowWrapper(
        "./__tests__/__fixtures__/TestComponent.tsx",
        "TestComponent"
      )();
      wrapper.mock("TestComponent");
      wrapper.unmock("TestComponent");

      const TestComponent = wrapper.TestComponent as Component;
      const { asFragment } = render(
        <TestComponent prop="lorem ipsum">children</TestComponent>
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
