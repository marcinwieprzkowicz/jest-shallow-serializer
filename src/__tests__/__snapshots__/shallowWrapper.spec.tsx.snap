// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shallowWrapper \`mock\` function lazy components does not render unresolved components 1`] = `
<DocumentFragment>
  <div
    data-shallow-name="TestComponentSuspense"
    data-shallow-props="{}"
  >
    <h1>
      Heading
    </h1>
  </div>
</DocumentFragment>
`;

exports[`shallowWrapper \`mock\` function lazy components renders resolved components 1`] = `
<DocumentFragment>
  <div
    data-shallow-name="TestComponentSuspense"
    data-shallow-props="{}"
  >
    <h1>
      Heading
    </h1>
    <div>
      I'm lazy
    </div>
  </div>
</DocumentFragment>
`;

exports[`shallowWrapper \`mock\` function mocks \`component\` 1`] = `
<DocumentFragment>
  <div
    data-shallow-name="TestComponent"
    data-shallow-props="{"prop":"lorem ipsum"}"
  >
    children
  </div>
</DocumentFragment>
`;

exports[`shallowWrapper \`mock\` function supports circular references 1`] = `
<DocumentFragment>
  <div
    data-shallow-name="TestComponent"
    data-shallow-props="{"prop":"lorem ipsum","obj":{"nested":{"arr":[{"item":{"value":"lorem ipsum"}}]},"circularReference":"[Circular]"}}"
  >
    children
  </div>
</DocumentFragment>
`;

exports[`shallowWrapper \`mock\` function supports functions 1`] = `
<DocumentFragment>
  <div
    data-shallow-name="TestComponent"
    data-shallow-props="{"prop":"lorem ipsum","obj":{"method":"[Function]","mockMethod":"[MockFunction]"}}"
  >
    children
  </div>
</DocumentFragment>
`;

exports[`shallowWrapper \`mock\` function uses \`options.renderAs\` as render tag 1`] = `
<DocumentFragment>
  <span
    data-shallow-name="TestComponent"
    data-shallow-props="{"prop":"lorem ipsum"}"
  >
    children
  </span>
</DocumentFragment>
`;

exports[`shallowWrapper \`mock\` function uses the first tag from the component body 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          data-shallow-name="TestComponentTd"
          data-shallow-props="{}"
        />
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`shallowWrapper \`unmock\` function unmocks \`component\` 1`] = `
<DocumentFragment>
  <div
    class="test-component"
    data-obj="[object Object]"
    data-prop="lorem ipsum"
  >
    <div
      class="test-component__body"
    >
      children
    </div>
  </div>
</DocumentFragment>
`;

exports[`shallowWrapper does not mock \`component\` by default 1`] = `
<DocumentFragment>
  <div
    class="test-component"
    data-obj="[object Object]"
    data-prop="lorem ipsum"
  >
    <div
      class="test-component__body"
    >
      children
    </div>
  </div>
</DocumentFragment>
`;
